pkg/integrity/dsse.go

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) 2022, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the LICENSE.md file
+// distributed with the sources of this project regarding your rights to use or distribute this
+// software.
+
+package integrity
+
+import (
+	"bytes"
+	"crypto"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/secure-systems-lab/go-securesystemslib/dsse"
+	"github.com/sigstore/sigstore/pkg/signature"
+	"github.com/sigstore/sigstore/pkg/signature/options"
+)
+
+var metadataMediaType = "application/vnd.sylabs.sif-metadata+json"
+
+type dsseEncoder struct {
+	es          *dsse.EnvelopeSigner
+	h           crypto.Hash
+	payloadType string
+}
+
+// newDSSEEncoder returns an encoder that signs messages in DSSE format according to opts, with key
+// material from ss. SHA256 is used as the hash algorithm, unless overridden by opts.
+func newDSSEEncoder(ss []signature.Signer, opts ...signature.SignOption) (*dsseEncoder, error) {
+	var so crypto.SignerOpts
+	for _, opt := range opts {
+		opt.ApplyCryptoSignerOpts(&so)
+	}
+
+	// If SignerOpts not explicitly supplied, set default hash algorithm.
+	if so == nil {
+		so = crypto.SHA256
+		opts = append(opts, options.WithCryptoSignerOpts(so))
+	}
+
+	dss := make([]dsse.SignVerifier, 0, len(ss))
+	for _, s := range ss {
+		ds, err := newDSSESigner(s, opts...)
+		if err != nil {
+			return nil, err
+		}
+
+		dss = append(dss, ds)
+	}
+
+	es, err := dsse.NewEnvelopeSigner(dss...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &dsseEncoder{
+		es:          es,
+		h:           so.HashFunc(),
+		payloadType: metadataMediaType,
+	}, nil
+}
+
+// signMessage signs the message from r in DSSE format, and writes the result to w. On success, the
+// hash function is returned.
+func (en *dsseEncoder) signMessage(w io.Writer, r io.Reader) (crypto.Hash, error) {
+	body, err := io.ReadAll(r)
+	if err != nil {
+		return 0, err
+	}
+
+	e, err := en.es.SignPayload(en.payloadType, body)
+	if err != nil {
+		return 0, err
+	}
+
+	return en.h, json.NewEncoder(w).Encode(e)
+}
+
+type dsseDecoder struct {
+	vs          []signature.Verifier
+	threshold   int
+	payloadType string
+}
+
+// newDSSEDecoder returns a decoder that verifies messages in DSSE format using key material from
+// vs.
+func newDSSEDecoder(vs ...signature.Verifier) *dsseDecoder {
+	return &dsseDecoder{
+		vs:          vs,
+		threshold:   1, // Envelope considered verified if at least one verifier succeeds.
```

Comment:
Is there ever a case when it'd be appropriate for this to be >1? Just trying to get my head around how you might use multiple verifiers with DSSE - but I think I know why it's 1 here.

Reply:
This is related to the fact that a DSSE envelope can have more than one signature in it. For example:

```
$ cat pkg/integrity/testdata/Test_dsseEncoder_signMessage/Multi.golden | jq
{
  "payloadType": "application/vnd.sylabs.sif-metadata+json",
  "payload": "eyJPbmUiOjEsIlR3byI6Mn0K",
  "signatures": [
    {
      "keyid": "SHA256:13cNV0A11AxgIQYCvlDmJlRi0CjRtjnYF47RzrMhlaU",
      "sig": "MEQCIFo8YCnASBqoUCa0T2H2L/crRYxZ9vFkSpbplBYo4IR6AiAUNulVHknRpFlrRWayZLY6vG9WIW5mx/ywLyI90KchQA=="
    },
    {
      "keyid": "SHA256:x6l8ZblpSSXGaPMCzySedWg88BwIFcz8jlPb6el0mFs",
      "sig": "SNnYRFIhDwWjk0pxoreaNiLea6L2WAFUm4boxnv7jiBNGmvMnbCxdsHYsTRBLXvMJHwEfKGvHFJmi9VvMe4JCQ=="
    },
    {
      "keyid": "SHA256:BhCwr7qZulYcOMSl2Jt2DuYHxHNnN6th4NdMqR/PGa4",
      "sig": "qtxC0N3TWRUOOF4nAFwf8izZMVhpGca/s0STBi2h/OU/lND9M4uPG70LMGJ+n2GhCOyKKLR5BpgtlUkBpwhsxiPDqyyXFE2/Rvu/MsNicNIal7A1E64X3iOrMmaXK7qHDY6TpwC0KlxTOsh2XHJSM/cItgebkiRn5ZaZl48/10IzMsq/nOr0k9fGdAdgeApnRAQzBuHzcSAMpz8k9ovbyecfwuNLxXk6PO3isetpFx2j1d11gNfmwE54lCQ9ZGC3hiTJVt9WLBP+xC5AGoiX9f5FQpRzQrg9xGjyfwZDF4PSE9UFfUAC4fGPdultxUPXp8afWocJwbDgZBOkUKgE2L16LtMYSPFMdmAy615Ah6AOyudDTY+6iUr8D7YFdXgkjuQOGxtk7Wh2AIwk1lTOF4nrpycNjOJawBW5AFxdjEJ0LvG/XEJgSC88RoAkQ0YdN7j5N8nNf4+bZJ+CmTXPWU0MdFVDgI59bJKUJU/lt1WM/ZEIzujCgtqYKwCc8LNl5Fruh+2nHmtsAS3bxxPv51Nbw5d8T316SBp0bhjY+R7OncQDaP2FQ+nwpUXuDX3Tr9pqMJxDgErbIATOdSaRQ3KB1iC5gzTwIikuwPIxAuB2Gb5wWGxhqqfx7iA38TpnP5x8YXsjGCseUxFjrKoj5uL1p6ayGXOPJy/D9FsQVtA="
    }
  ]
}
```

The `threshold` is related to the cardinality of the set of verifiers that were able to verify the signature. In other words, how many `Verifier`s were involved in verifying the signature. The simple (and usual case) will be one signature per envelope, which is easy to understand. In the more complicated case where there are multiple signatures, it's really a question of semantics. Successfully verifying could be interpreted in a variety of ways:

- All signatures were verified by different keys. In this case, you could set the threshold to the number of signatures.
- All signatures were verified, but at least one key was used more than once. In this case, you can't set the threshold to the number of signatures. If you set the threshold to the number of verifiers that were passed, it works, as long as every signature is signed at least once by every verifier that was passed. This definitely isn't the current semantics.
- Let the caller decide the semantics. The caller passes in various key material, and examines the [VerifyResult](https://pkg.go.dev/github.com/sylabs/sif/v2/pkg/integrity#VerifyResult) for each signature. I intend to expose a `.Keys` method that will return the public key(s) that successfully validated the signature.

Let me know if that makes sense!

Reply:
👍